<?xml version="1.0"?>
<launch>
    <arg name="publish_images" default="true"/>
    <arg name="jsdev" default="/dev/input/js0"/>
    <rosparam file="$(find pitakuru)/config/bright_env/pitakuru.yaml" command="load" ns="pitakuru" />

 <!-- スイッチの入出力制御 -->
    
    <node pkg="pitakuru" type="switch.py" name="switch">
        <rosparam command="load" file="$(find pitakuru)/config/bright_env/switch.yaml" />
    </node>

    <!-- ステートマシン (状態遷移) -->
    <node pkg="pitakuru" type="states.py" name="states"/>
    <!-- カメラから画像を読み込んで配信 -->
 <node name="uvc_camera_node" pkg="uvc_camera" type="uvc_camera_node" output="screen"> -->
    <!-- <param name="device" type="string" value="/dev/input/by-id/usb-KYE_Systems_Corp._USB_Camera_200901010001-event-if00"/> -->
    <param name="device" type="string" value="/dev/video0"/>
    <param name="width" type="int" value="640"/> <!-- 800 --> <!-- if you move image size you also need to modify cyaml -->
    <param name="height" type="int" value="480"/> <!-- 600 -->
    <param name="fps" type="int" value="7"/><!-- 24 -->
    <param name="camera_info_url" value="file://$(find pitakuru)/config/bright_env/aruco/camera640.yaml"/><!-- camera600.yaml -->
    </node>

    <!-- ArUco検知 -->
 <!--    <node pkg="aruco_detect" type="aruco_detect" name="aruco_detect">
        <rosparam command="load" file="$(find pitakuru)/config/bright_env/aruco/aruco.yaml"/>
        <param name="publish_images" value="$(arg publish_images)"/>
        <remap from="camera" to="/image_raw"/>
    </node>  -->  <!--<remap from="camera" to="/usb_cam/image_raw"/>-->

<node pkg="pitakuru" type="controller.py" name="controller"/>

<!-- launch file for ps4 controller -->
<include file="$(find joy)/launch/joytestps4.launch"/> 

<!-- conversion needed from quaternions to euler needed for control_xy node-->
<node pkg="conversion" type="quatCo.py" name="quatCo" output="screen">	
</node>

<!-- conversion needed from quaternions to euler needed for control_xy node-->	
<node pkg="conversion" type="quatCo2.py" name="quatCo2" output="screen"> 
</node>


  
    <!-- 速度指令変換 -->
    <rosparam command="load" file="$(find pitakuru)/config/bright_env/movement.yaml" />
    <node pkg="pitakuru" type="movement.py" name="movement" />
    <node pkg="pitakuru" type="karugamo.py" name="karugamo" />
    <node pkg="pitakuru" type="manual.py" name="manual" />
    
    <!-- 音声再生 -->
    
    <node pkg="pitakuru" type="sound.py" name="sound" />

    <!-- package for configure obstacle detector node -->
 
    <include file="$(find obstacle_detector)/launch/test_experim.launch"/>  
 
    <!-- opencv detect people package -->
<!-- <node pkg="pitakuru" type="people_detect.py" name="people_detect"/>  -->

    <include file="$(find darknet_ros)/launch/yolo_v3.launch"/> 
    <include file="$(find sort_track)/launch/sort_big.launch"/> 

    <node pkg="peop_extract" type="peop_ext_sort.py" name="peop_ext_yolo"/>

    <node pkg="pitakuru" type="lidar_detect.py" name="lidar_detect_f" />

<!-- keiganモーター制御/オドメトリ算出ノード -->
<node pkg="pypro" name="pitakuru_wheels" type="pitakuruWheels.py">
<param name="max_speed"    type="double"	value="3500"/>  <!-- max speed tah can be sent to motors-->
</node>

<node pkg="pitakuru" name="pitakuru_lidar_detect_fail" type="lidar_detect.py"/>



<!-- extract obstacles form region of interest and send to control_xy -->
 
<node pkg="obs_extract" type="main_sorted_kalman.py" name="obj_track" >
<param name="front_detection"    type="double"	value="7"/> 
<param name="side_detection"    type="double"	value="2.5"/>  
<param name="radius_follow"    type="double"	value="0.37"/>  
</node>


<node pkg="costmap" type="main.py" name="cost_detect">
<param name="dilation"    type="int"	value="4"/> <!-- 3.5 --><!-- 5 -->
</node>
<!--  
<node pkg="tf" type="static_transform_publisher" name="base_link_map_broadcaster" args="0 0 0 0 0 0 map base_link 50" />
-->
 <!--
<node pkg="tf" type="static_transform_publisher" name="bases_broadcaster" args="0 0 0 0 0 0 base_link base_link3 100" />
 -->

<node name="map_to_odom_base" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 0 map odom"/>
<!--
<node pkg="pitakuru" type="imu.py" name="imuw"/>
<node pkg="tf" type="static_transform_publisher" name="imu_base_link_broadcaster" args="0.1 0 0.2 0 0 0 base_link immu 100" />
-->
<!-- localization -->
<!-- 
<node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true">
    <rosparam command="load" file="$(find robot_localization)/params/ekf_template_test2.yaml" />
  </node>
-->

<!-- navigation -->

<include file="$(find pitakuru)/launch/navigation.launch"/>

<node name="frame_base" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 0 map my_frame"/>

<!-- rosbag nodes -->
<arg name="record_flag" default="false" /> <!-- set to true if i want to record-->

<node pkg="rosbag" type="record" name="rosbag_record"
 args = "record -o /home/xavier/floornew.bag /alerts /camera_info /diagnostics /euler /fiducial_vertices /image_raw /j0/joy  /movement/karugamo/joy /movement/manual/joy /obstacles /pcl /peopAng /peopDist /raw_obstacles /rear_scan /rosout /rosout_agg /scan /state /tf /tf_static /trigger_action /peopAng2 /peopAngSpeed /path /pathpeop /peop_way /peop_obst_id /peopAngSpeed /visualization_marker /peop_ang_yolo /peop_obst_id /sorted_tracked /stateWheels /pitakuru_states /tracked"  
   if = "$(arg record_flag)" />

<arg name="record_flag2" default="false" /> <!-- set to true if i want to record-->

<node pkg="rosbag" type="record" name="rosbag_record"
 args = "record -o /home/xavier/sea.bag  /scan "  
   if = "$(arg record_flag2)" />

<arg name="record_load_flag" default="false" /> <!-- set to true if i want to record-->

<node pkg="rosbag" type="record" name="rosbag_record"
 args = "record -o /home/xavier/test.bag /stateWheels /alerts /state"  
   if = "$(arg record_load_flag)" />

<arg name="record_m_flag" default="false" /> <!-- set to true if i want to record-->

<node pkg="rosbag" type="record" name="rosbag_record"
 args = "record -o /home/xavier/test_scan.bag /stateWheels /alerts /state /m1_deriv /m2_deriv"  
   if = "$(arg record_m_flag)" />


<!--<include file="$(find hector_slam_launch)/launch/tutorial.launch"/> -->
 
<!-- SLAM地図作成ノード -->
<!--

    <param name="use_sim_time" value="false"/>
    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" clear_params="true">
      <rosparam command="load" file="$(find pitakuru)/config/gmapping.yaml" />
      <remap from="scan" to="/scan" />
      <param name="map_frame" value="map"/>
      <param name="base_frame" value="base_link" />
      <param name="odom_frame" value="odom" />
    </node>
-->

<!--
<node pkg="pitakuru" type="imu.py" name="imuw"/>
<node pkg="tf" type="static_transform_publisher" name="imu_base_link_broadcaster" args="0.1 0 0.2 0 0 0 base_link immu 100" />
-->

 #### map ##############################


<arg name="map_file" default="$(find pitakuru)/maps/floor6_hokuyo.yaml"/>
<node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" />



  #### start the laser scan_matcher ##############################
<!--
  <node pkg="laser_scan_matcher" type="laser_scan_matcher_node" 
    name="laser_scan_matcher_node" output="screen">
    <param name="use_vel"        value="true"/>
    <param name="max_iterations" value="10"/>
  </node>
 -->
  #### start the alpha-beta filter ###############################
<!-- 
  <node pkg="ab_filter" type="ab_filter_pose2d_node" 
    name="ab_filter_pose2d" output="screen">
    <remap from="mav/pose2D"      to="pose_2D"/>
    <remap from="mav/pose2D_f"    to="pose2D_f"/>
    <remap from="mav/pose2D_unf"  to="pose2D_unf"/>
    <param name="alpha" value="1.0"/>
    <param name="beta"  value="0.5"/>
    <param name="publish_unfiltered"  value="true"/>
  </node>
-->

</launch>
